quesma/queryparser/pancake_transformer.go

```diff
@@ -292,6 +297,57 @@ func (a *pancakeTransformer) findParentBucketLayer(layers []*pancakeModelLayer,
 	return layer, nil
 }
 
+func (a *pancakeTransformer) createTopHitPancakes(pancake *pancakeModel) (result []*pancakeModel, err error) {
+	for layerIdx, layer := range pancake.layers {
+		metricsWithoutTopHits := make([]*pancakeModelMetricAggregation, 0, len(layer.currentMetricAggregations))
+		for _, metric := range layer.currentMetricAggregations {
+			if _, isTopHits := metric.queryType.(metrics_aggregations.TopHits); isTopHits {
+				canOptimize := layerIdx == len(pancake.layers)-1 && len(layer.currentMetricAggregations) == 1
+				for _, layer2 := range pancake.layers[:layerIdx] {
+					if len(layer2.currentMetricAggregations) > 0 {
+						canOptimize = false
+					}
+					if layer2.nextBucketAggregation != nil {
+						switch layer2.nextBucketAggregation.queryType.(type) {
+						case bucket_aggregations.CombinatorAggregationInterface:
+							// TODO: possible to implement, by generating more queries, skipped as it is rare
+							return nil, fmt.Errorf("top_hits can't be in filter(s)/range/dataRange aggregation")
+						}
+					}
+				}
+
+				if canOptimize { // if this is just one top_hits with buckets than we can optimize
+					// and don't need to create new pancake
+					metricsWithoutTopHits = append(metricsWithoutTopHits, metric)
+				} else {
+					newLayers := make([]*pancakeModelLayer, layerIdx)
+					for i := range newLayers {
+						newLayers[i] = &pancakeModelLayer{
+							currentMetricAggregations: make([]*pancakeModelMetricAggregation, 0),
+							nextBucketAggregation:     pancake.layers[i].nextBucketAggregation,
+						}
+					}
+					newLayers = append(newLayers, &pancakeModelLayer{
+						currentMetricAggregations: []*pancakeModelMetricAggregation{metric},
+						nextBucketAggregation:     nil,
+					})
+
+					newPancake := pancakeModel{
+						layers:      newLayers,
+						whereClause: pancake.whereClause,
+						sampleLimit: pancake.sampleLimit,
+					}
+					result = append(result, &newPancake)
+				}
+			} else {
+				metricsWithoutTopHits = append(metricsWithoutTopHits, metric)
+			}
```

Comment:
```suggestion
			_, isTopHits := metric.queryType.(metrics_aggregations.TopHits)
if !isTopHits {
	metricsWithoutTopHits = append(metricsWithoutTopHits, metric)
	continue
}

canOptimize := layerIdx == len(pancake.layers)-1 && len(layer.currentMetricAggregations) == 1
for _, layer2 := range pancake.layers[:layerIdx] {
	if len(layer2.currentMetricAggregations) > 0 {
		canOptimize = false
	}
	if layer2.nextBucketAggregation != nil {
		if _, isCombinator := layer2.nextBucketAggregation.queryType.(bucket_aggregations.CombinatorAggregationInterface); isCombinator {
			// TODO: possible to implement, by generating more queries, skipped as it is rare
			return nil, fmt.Errorf("top_hits can't be in filter(s)/range/dataRange aggregation")
		}
	}
}

if canOptimize { // if this is just one top_hits with buckets than we can optimize
	// and don't need to create new pancake
	metricsWithoutTopHits = append(metricsWithoutTopHits, metric)
} else {
	newLayers := make([]*pancakeModelLayer, layerIdx)
	for i := range newLayers {
		newLayers[i] = newPancakeModelLayer()
		newLayers[i].nextBucketAggregation = pancake.layers[i].nextBucketAggregation
	}
	
	topHitsLayer := newPancakeModelLayer()
	topHitsLayer.currentMetricAggregations = append(topHitsLayer.currentMetricAggregations, metric)
	newLayers = append(newLayers, topHitsLayer)

	newPancake := pancakeModel{
		layers:      newLayers,
		whereClause: pancake.whereClause,
		sampleLimit: pancake.sampleLimit,
	}
	result = append(result, &newPancake)
}
```

Reply:
This get's modified in subsequent `top_metrics` PR. Let's apply changes there.